core/src/main/scala/akka/kafka/internal/ProducerStage.scala

```diff
@@ -6,126 +6,284 @@
+    override def postStop() = {
+      log.debug("Stage completed")
 
-              if (awaitingConfirmation.decrementAndGet() == 0 && inIsClosed)
-                checkForCompletionCB.invoke(())
-            }
-          })
-          awaitingConfirmation.incrementAndGet()
-          push(out, r.future)
+      if (closeProducerOnStop) {
+        try {
+          // we do not have to check if producer was already closed in send-callback as `flush()` and `close()` are effectively no-ops in this case
+          producer.flush()
+          producer.close(closeTimeout.toMillis, TimeUnit.MILLISECONDS)
+          log.debug("Producer closed")
         }
-
-        override def onUpstreamFinish() = {
-          inIsClosed = true
-          completionState = Some(Success(()))
-          checkForCompletion()
+        catch {
+          case NonFatal(ex) => log.error(ex, "Problem occurred during producer close")
         }
+      }
+
+      super.postStop()
+    }
+  }
+
+  /**
+   * Transactional (Exactly-Once) Producer State Logic
+   */
+  class TransactionProducerStageLogic(shape: Shape, producer: Producer[K, V], inheritedAttributes: Attributes, commitIntervalMs: Long) extends ProducerStageLogic(shape, producer, inheritedAttributes) with StageLogging with MessageCallback[K, V, P] with ProducerCompletionState {
+    private val commitSchedulerKey = "commit"
+    private val messageDrainIntervalMs = 10
 
-        override def onUpstreamFailure(ex: Throwable) = {
-          inIsClosed = true
-          completionState = Some(Failure(ex))
-          checkForCompletion()
+    private var isTransactionOpen = false
+    private var commitInProgress = false
+    private val batchOffsets = new TransactionOffsetBatch
+
+    override def preStart(): Unit = {
+      initTransactions()
+      beginTransaction()
+      scheduleOnce(commitSchedulerKey, commitIntervalMs.milliseconds)
+      super.preStart()
+    }
+
+    setHandler(out, new OutHandler {
+      override def onPull() = {
+        // stop pulling while a commit is in process so we can drain any outstanding message acknowledgements
+        if (!commitInProgress) {
+          tryPull(in)
         }
-      })
+      }
+    })
 
-      override def postStop() = {
-        log.debug("Stage completed")
+    override protected def onTimer(timerKey: Any): Unit =
+      if (timerKey == commitSchedulerKey) {
+        onCommitInterval()
+      }
 
-        if (closeProducerOnStop) {
-          try {
-            // we do not have to check if producer was already closed in send-callback as `flush()` and `close()` are effectively no-ops in this case
-            producer.flush()
-            producer.close(closeTimeout.toMillis, TimeUnit.MILLISECONDS)
-            log.debug("Producer closed")
-          }
-          catch {
-            case NonFatal(ex) => log.error(ex, "Problem occurred during producer close")
-          }
+    private def onCommitInterval(): Unit = {
+      if (!commitInProgress) {
```

Comment:
`onCommitInterval` is called on the normal commit interval and when draining producer ack's.  The guard is just to prevent re-assigning `commitInProgress = true` in the latter case, but it doesn't really matter since its threadsafe anyway.  I could have created a separate schedule, but since the callback is the same (except to assign `commitInProgress`) I just re-used it.